### DISCUSSION TITLE
1.2.0 How do we run the sql create tables code before first running server

### DISCUSSION THREAD

**Discussion by jason-downs** [serverpod/serverpod#1796] [General]
I do not have data to migrate and haven't even run the server code yet on this particular project before upgrading to 1.2.0. So not sure if this is a "migration" issue or what, but...

Every time I try to run my server bin/main.dart I get the "relation not found" error for the database. I followed all the steps regarding migrating with no data to preserve, and I can see the create table code for this table in my definition.sql and migration.sql files so somehow it isn't getting run?

In the older version docs I recall having to run something to initialize the db first, but this no longer seems to be there and I assume has been replaced with the migration functionality? (I am runnnig the --apply-migration flag to no avail).

Also, the old(?) sql files in the 'generated' folders - can these all be safely deleted? Including the ones that dealt with serverpod auth? The documentation is not clear about this. Perhaps this is causing my problem?

**Comment by jason-downs** [comment]


**Reply by Isakdl** [reply]
You have to run a migration (all this does is execute any migration that is not applied to the database).

`dart run bin/main.dart --apply-migration`

This assumes you have a migration in your project. If you don't you can create one with `serverpod create-migration` from your server project. 

> Also, the old(?) sql files in the 'generated' folders - can these all be safely deleted? Including the ones that dealt with serverpod auth? The documentation is not clear about this. Perhaps this is causing my problem?

Yes they can safely be deleted, they will no longer be update and/or generated. Including the auth module, modules has full support in the migration system. 

I think your issue is that you do not have a migration in your project, if you do there should be a `migration` directory in the root of the server project.

See the docs for more info: https://docs.serverpod.dev/concepts/database/migrations

**Comment by jason-downs** [comment]
There is a migration folder that I had already created with `serverpod create-migration`, and I was indeed running `dart run bin/main.dart --apply-migration` when I got the error. 

What eventually worked for me was I just copied all the SQL in the 'definitions' file in the migration folder and ran it in Postico, then was able to run `dart run bin/main.dart --apply-migration` without this problem. (Although I did get another problem related to 'no deserializer found' that I'm trying to debug now.)

**Reply by Isakdl** [reply]
That sounds strange, do you have any manually defined tables? If I remember the stack trace you posted correctly you got a mismatch error when quering the db? Serverpod should do an integrity check on boot and give you a warning if there is any mismatch between the db schema and the generated code.

**Reply by jason-downs** [reply]
Here's the error stack trace again. By manually defined tables, does this mean my own custom SQL? If so then no there is none of that, everything is the auto-generated code based on the classes definied in my protocol yaml files.

It's worth noting that the relation "regular_inflection" mentioned here is from my external serverpod module that I've imported into this project via pubspec. In that module, I also ran `serverpod create-migration` before importing into my main project. I don't know if that has anything to do with the error if this regular_inflection from the module just happens to be the first that it encounters as the server tries to start up.

```
SERVERPOD version: 1.2.1, dart: 3.2.5 (stable) (Tue Jan 16 15:02:13 2024 +0000) on "macos_arm64", time: 2024-01-24 15:05:05.768504Z
mode: development, role: monolith, logging: normal, serverId: default
Unhandled exception:
Exception: Unable to initialize object graph:
Exception: Unable to initialize Lingbat language pack databases:
PostgreSQLSeverity.error 42P01: relation "regular_inflection" does not exist 
package:postgres/src/connection.dart 520:18                                                _PostgreSQLExecutionContextMixin._query
package:postgres/src/connection.dart 481:7                                                 _PostgreSQLExecutionContextMixin.query
package:postgres/src/connection.dart 549:22                                                _PostgreSQLExecutionContextMixin.mappedResultsQuery
package:postgres_pool/postgres_pool.dart 820:23                                            _PgExecutionContextWrapper.mappedResultsQuery.<fn>
package:postgres_pool/postgres_pool.dart 746:27                                            _PgExecutionContextWrapper._run
package:postgres_pool/postgres_pool.dart 819:12                                            _PgExecutionContextWrapper.mappedResultsQuery
package:postgres_pool/postgres_pool.dart 661:16                                            PgPool.mappedResultsQuery.<fn>
package:postgres_pool/postgres_pool.dart 352:22                                            PgPool.run.<fn>.<fn>
package:postgres_pool/postgres_pool.dart 451:27                                            PgPool._useOrCreate
package:postgres_pool/postgres_pool.dart 423:14                                            PgPool._withConnection.<fn>
package:executor/src/executor_impl.dart 61:19                                              _Executor.scheduleTask
package:postgres_pool/postgres_pool.dart 351:18                                            PgPool.run.<fn>
package:retry/retry.dart 131:16                                                            RetryOptions.retry
package:postgres_pool/postgres_pool.dart 349:14                                            PgPool.run
package:serverpod/src/database/database_connection.dart 394:20                             DatabaseConnection.mappedResultsQuery
package:serverpod/src/database/database_connection.dart 422:18                             DatabaseConnection._deserializedMappedQuery
package:lingbat_server/src/code/data/serverpod_pgsql/svpd_pgsql_inflections_db.dart 22:32  SvpdPgsqlInflectionsDatabase.init
package:lingbat_server/src/code/logic/startup/database_loader.dart 60:29                   DatabaseLoader.loadTargetLanguagePack
package:lingbat_server/src/code/logic/startup/object_graph_constructor.dart 60:36          ObjectGraphConstructor.createObjectGraph
package:lingbat_server/server.dart 79:26                                                   run
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 351:18                                            PgPool.run.<fn>
package:retry/retry.dart 131:16                                                            RetryOptions.retry
package:postgres_pool/postgres_pool.dart 349:14                                            PgPool.run
package:serverpod/src/database/database_connection.dart 394:20                             DatabaseConnection.mappedResultsQuery
package:serverpod/src/database/database_connection.dart 422:18                             DatabaseConnection._deserializedMappedQuery
package:lingbat_server/src/code/data/serverpod_pgsql/svpd_pgsql_inflections_db.dart 22:32  SvpdPgsqlInflectionsDatabase.init
package:lingbat_server/src/code/logic/startup/database_loader.dart 60:29                   DatabaseLoader.loadTargetLanguagePack
package:lingbat_server/src/code/logic/startup/object_graph_constructor.dart 60:36          ObjectGraphConstructor.createObjectGraph
package:lingbat_server/server.dart 79:26                                                   run

#0      DatabaseLoader.loadTargetLanguagePack (package:lingbat_server/src/code/logic/startup/database_loader.dart:86:7)
<asynchronous suspension>
#1      ObjectGraphConstructor.createObjectGraph (package:lingbat_server/src/code/logic/startup/object_graph_constructor.dart:60:36)
<asynchronous suspension>
#2      run (package:lingbat_server/server.dart:79:26)
<asynchronous suspension>

#0      ObjectGraphConstructor.createObjectGraph (package:lingbat_server/src/code/logic/startup/object_graph_constructor.dart:259:7)
<asynchronous suspension>
#1      run (package:lingbat_server/server.dart:79:26)
<asynchronous suspension>
```

**Reply by jason-downs** [reply]
(Note the above error is the result of running `dart run bin/main.dart --apply-migrations` in my project's server directory. 

The auto-generated SQL in definition.sql does contain this regular_inflection that it claims can't be found:
```
-- Class RegularInflection as table regular_inflection
--
CREATE TABLE "regular_inflection" (
```
(...not including the table details to hide proprietary info)

**Reply by jason-downs** [reply]
But as mentioned earlier, if I do copy the SQL from definition.sql and execute that SQL code, then run `dart run bin/main.dart --apply-migrations` the server will start up without error. It's just that the docs imply that with version 1.2.0 it's no longer necessary to manually copy and execute the SQL code to create the tables, and that the migration functionality should do that instead.

**Reply by jason-downs** [reply]
Also note that this happens when the db is brand-new with no existing tables or data to be populated yet.

**Reply by Isakdl** [reply]
Thank you for the added context! Yes this should work without having to execute manually, having a custom module does complicate things a little. You will have to create the migration in the module first and then another migration in your project. 

Just to make sure, which definition.sql are you executing? The one created in your project or the one in the module? 
Does the definition.sql in your project contain the table definitions from your module? 

If not could you verify that the `config/generator.yaml` in the module has the type set to `module`.

This is what the file looks like in the auth module as an example:

```yaml
type: module
nickname: auth

client_package_path: ../serverpod_auth_client
```

One last thought just to make sure, in your server.dart are you executing any queries before you call `start`? I.E. this line:

```dart
// Start the server
  await pod.start();
```

The start method is the one doing the migration and other booting sequences of the server. Meaning if you are trying to read from the database before this method is called no tables will be available.

**Reply by jason-downs** [reply]
Oh -- that's what it is! All this database-related code I'm running on startup is before the `pod.start()`!

**Reply by jason-downs** [reply]
I can probably move all that to after the pod starts without any trouble, hopefully then it will fix things. For some reason I had this mentality of all the preparation work I needed to do upon startup should be before that line but there's no reason it needs to be that way.

**Reply by Isakdl** [reply]
Perfect! 😄
